bitcoin/src/blockdata/transaction.rs

```diff
@@ -987,6 +987,51 @@ impl Transaction {
         // `Transaction` docs for full explanation).
         self.input.is_empty()
     }
+
+    /// Returns a reference to the input at `input_index` if it exists.
+    #[inline]
+    pub fn txin(&self, input_index: usize) -> Result<&TxIn, IndexOutOfBoundsError> {
+        if input_index >= self.input.len() {
+            return Err(IndexOutOfBoundsError {
+                index: input_index,
+                length: self.input.len(),
+            });
+        }
+        Ok(&self.input[input_index])
+    }
+
+    /// Returns a reference to the output at `output_index` if it exists.
+    #[inline]
+    pub fn txout(&self, output_index: usize) -> Result<&TxOut, IndexOutOfBoundsError> {
+        if output_index >= self.output.len() {
+            return Err(IndexOutOfBoundsError {
+                index: output_index,
+                length: self.output.len(),
+            });
+        }
+        Ok(&self.output[output_index])
+    }
+}
+
+/// Error when attempting to do an out of bounds access on transaction inputs or outputs.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct IndexOutOfBoundsError {
+    /// Attempted index access.
+    pub index: usize,
+    /// Length of the vector where access was attempted.
+    pub length: usize,
+}
```

Comment:
This is fine, especially with `non_exhaustive` and I think we can also commit to these fields but it might be really cool to also provide txid so that if one runs a program processing many transactions the problematic one it can be easily found.

However hashing may be a bit expensive even for error path, so maybe this should be behind `fancy_errors` feature or something.

Also we could carry the information whether it was inputs or outputs.

Reply:
I don't want to introduce a new feature in this PR, I agree we should consider having one but there that smells like there will be a bit of back-and-forth on that.

I'm not totally happy with the.